garrysmod/lua/vgui/dscrollpanel.lua

```diff
@@ -27,119 +30,106 @@ function PANEL:Init()
 	self:SetPaintBackgroundEnabled( false )
 	self:SetPaintBorderEnabled( false )
 	self:SetPaintBackground( false )
-
 end
 
 function PANEL:AddItem( pnl )
-
 	pnl:SetParent( self:GetCanvas() )
-
 end
 
 function PANEL:OnChildAdded( child )
-
 	self:AddItem( child )
-
 end
 
 function PANEL:SizeToContents()
-
-	self:SetSize( self.pnlCanvas:GetSize() )
-
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		self:SetSize( canvas:GetSize() )
+	end
 end
 
 function PANEL:GetVBar()
-
 	return self.VBar
-
 end
 
 function PANEL:GetCanvas()
-
 	return self.pnlCanvas
-
 end
 
 function PANEL:InnerWidth()
-
 	return self:GetCanvas():GetWide()
-
-end
-
-function PANEL:Rebuild()
-
-	self:GetCanvas():SizeToChildren( false, true )
-
-	-- Although this behaviour isn't exactly implied, center vertically too
-	if ( self.m_bNoSizing && self:GetCanvas():GetTall() < self:GetTall() ) then
-
-		self:GetCanvas():SetPos( 0, ( self:GetTall() - self:GetCanvas():GetTall() ) * 0.5 )
-
-	end
-
 end
 
 function PANEL:OnMouseWheeled( dlta )
-
-	return self.VBar:OnMouseWheeled( dlta )
-
+	local vbar = self:GetVBar()
+	if IsValid( vbar ) then
+		return vbar:OnMouseWheeled( dlta )
+	end
 end
 
 function PANEL:OnVScroll( iOffset )
-
-	self.pnlCanvas:SetPos( 0, iOffset )
-
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		canvas:SetPos( 0, iOffset )
+	end
 end
 
 function PANEL:ScrollToChild( panel )
-
 	self:InvalidateLayout( true )
 
-	local x, y = self.pnlCanvas:GetChildPosition( panel )
-	local w, h = panel:GetSize()
-
-	y = y + h * 0.5
-	y = y - self:GetTall() * 0.5
-
-	self.VBar:AnimateTo( y, 0.5, 0, 0.5 )
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		local vbar = self:GetVBar()
+		if IsValid( vbar ) then
+			local _, y = canvas:GetChildPosition( panel )
+			vbar:AnimateTo( (y + panel:GetTall() / 2) - self:GetTall() / 2, 0.5, 0, 0.5 )
+		end
+	end
+end
 
+function PANEL:Rebuild()
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		canvas:SizeToChildren( false, true )
+
+		-- Although this behaviour isn't exactly implied, center vertically too
+		local canvasHeight = canvas:GetTall()
+		if self.m_bNoSizing and (canvasHeight < self:GetTall()) then
+			canvas:SetPos( 0, (self:GetTall() - canvasHeight) / 2 )
+		end
+	end
 end
 
--- Avoid an infinite loop
 function PANEL:PerformLayoutInternal()
-
-	local Tall = self.pnlCanvas:GetTall()
-	local Wide = self:GetWide()
-	local YPos = 0
-
-	self:Rebuild()
-
-	self.VBar:SetUp( self:GetTall(), self.pnlCanvas:GetTall() )
-	YPos = self.VBar:GetOffset()
-
-	if ( self.VBar.Enabled ) then Wide = Wide - self.VBar:GetWide() end
-
-	self.pnlCanvas:SetPos( 0, YPos )
-	self.pnlCanvas:SetWide( Wide )
-
-	self:Rebuild()
-
-	if ( Tall != self.pnlCanvas:GetTall() ) then

```

Comment:
This was removed for seemingly no good reason?

Reply:
in my logic it makes no sense, it is already rebuilded above

Reply:
The comment is not about the Rebuild call, but regarding that call, if the size of the canvas has changed - you gotta rebuild because the size and position of everything on the canvas may have changed, which is in fact what is causing the "infinite loop" problem you are trying to solve.

It's not as simple as removing that function call.